server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ChangelogAction.java

```diff
@@ -19,152 +19,101 @@
  */
 package org.sonar.server.qualityprofile.ws;
 
-import com.google.common.base.Optional;
 import java.util.Date;
-import java.util.Map.Entry;
-import org.elasticsearch.action.search.SearchResponse;
-import org.elasticsearch.search.SearchHit;
-import org.sonar.api.resources.Languages;
+import java.util.Map;
 import org.sonar.api.server.ws.Request;
 import org.sonar.api.server.ws.Response;
 import org.sonar.api.server.ws.WebService.NewAction;
 import org.sonar.api.server.ws.WebService.NewController;
 import org.sonar.api.server.ws.WebService.Param;
 import org.sonar.api.utils.DateUtils;
-import org.sonar.api.utils.Paging;
 import org.sonar.api.utils.text.JsonWriter;
-import org.sonar.db.DbClient;
-import org.sonar.db.DbSession;
-import org.sonar.db.rule.RuleDto;
-import org.sonar.db.user.UserDto;
-import org.sonar.server.activity.index.ActivityIndex;
-import org.sonar.server.es.SearchOptions;
-import org.sonar.server.exceptions.NotFoundException;
-import org.sonar.server.qualityprofile.QProfileActivity;
-import org.sonar.server.qualityprofile.QProfileActivityQuery;
-import org.sonar.server.qualityprofile.QProfileFactory;
-import org.sonar.server.search.Result;
-
-import static org.sonar.api.utils.Paging.forPageIndex;
+import org.sonar.db.qualityprofile.QProfileChangeQuery;
+import org.sonar.db.qualityprofile.QualityProfileDto;
+
 import static org.sonar.server.es.SearchOptions.MAX_LIMIT;
 
 public class ChangelogAction implements QProfileWsAction {
 
   private static final String PARAM_SINCE = "since";
   private static final String PARAM_TO = "to";
 
-  private DbClient dbClient;
-  private ActivityIndex activityIndex;
-  private QProfileFactory profileFactory;
-  private Languages languages;
+  private final ChangelogLoader changelogLoader;
+  private final QProfileFinder profileFinder;
 
-  public ChangelogAction(DbClient dbClient, ActivityIndex activityIndex, QProfileFactory profileFactory, Languages languages) {
-    this.dbClient = dbClient;
-    this.activityIndex = activityIndex;
-    this.profileFactory = profileFactory;
-    this.languages = languages;
+  public ChangelogAction(ChangelogLoader changelogLoader, QProfileFinder profileFinder) {
+    this.changelogLoader = changelogLoader;
+    this.profileFinder = profileFinder;
   }
 
   @Override
   public void define(NewController context) {
-    NewAction changelog = context.createAction("changelog")
+    NewAction wsAction = context.createAction("changelog")
       .setSince("5.2")
       .setDescription("Get the history of changes on a quality profile: rule activation/deactivation, change in parameters/severity. " +
         "Events are ordered by date in descending order (most recent first).")
       .setHandler(this)
       .setResponseExample(getClass().getResource("example-changelog.json"));
 
-    QProfileIdentificationParamUtils.defineProfileParams(changelog, languages);
+    profileFinder.defineProfileParams(wsAction);
```

Comment:
I didn't propagate the call everywhere yet because I want to discuss about that with Teryk first.